typescript/src/redemption.ts

```diff
@@ -0,0 +1,291 @@
+// @ts-ignore
+import bcoin from "bcoin"
+// @ts-ignore
+import hash160 from "bcrypto/lib/hash160"
+import { BigNumber } from "ethers"
+import {
+  createKeyRing,
+  RawTransaction,
+  UnspentTransactionOutput,
+  Client as BitcoinClient,
+} from "./bitcoin"
+import { Bridge } from "./bridge"
+
+/**
+ * Represents a redemption request.
+ */
+export interface RedemptionRequest {
+  /**
+   * The address of the redeemer.
+   */
+  redeemerAddress: string
+
+  /**
+   * The amount of Bitcoins in satoshis that is requested to be redeemed.
+   * The actual value of the output in the Bitcoin transaction will be decreased
+   * by the sum of the fee share and the treasury fee for this particular output.
+   */
+  requestedAmount: BigNumber
+
+  /**
+   * The amount of Bitcoins in satoshis that is subtracted from the amount of
+   * the redemption request and used to pay the treasury fee.
+   * The value should be exactly equal to the value of treasury fee in the Bridge
+   * on-chain contract at the time the redemption request was made.
+   */
+  treasuryFee: BigNumber
+
+  /**
+   * The amount of Bitcoins in satoshis that is subtracted from the amount of
+   * the redemption request and used to pay the transaction network fee.
+   * The value must not exceed the value of the Bridge's txMaxFee parameter
+   * that was in force at the moment the redemption request was made. The sum
+   * of the txFee values taken from all redemption outputs is equivalent to
+   * the total network fee of the entire redemption transaction.
+   */
+  txFee: BigNumber
+
+  /**
+   * UNIX timestamp the request was created at.
+   */
+  requestedAt: number
+}
+
+/**
+ * Handles pending redemption requests by creating a redemption transaction
+ * transferring Bitcoins from the wallet's main UTXO to the redeemer addresses
+ * and broadcasting it. The change UTXO resulting from the transaction becomes
+ * the new main UTXO of the wallet.
+ * @dev It is up to the caller to ensure the wallet key and each of the redeemer
+ *      addresses represent a valid pending redemption request in the Bridge.
+ *      If this is not the case, an exception will be thrown.
+ * @param bitcoinClient - The Bitcoin client used to interact with the network
+ * @param bridge - The handle to the Bridge on-chain contract
+ * @param walletPrivateKey = The private kay of the wallet in the WIF format
+ * @param mainUtxo - The main UTXO of the wallet. Must match the main UTXO
+ *        held by the on-chain Bridge contract.
+ * @param redeemerAddresses - The list of redeemer addresses
```

Comment:
Step 3: I think the redeemer's Bitcoin address should be used only when the redeemer requests the redemption. From that point, we should use `redeemerOutputScript` as the address doesn't matter for the wallet and the `Bridge`. Hence, I propose to rename `redeemerAddresses` to `redeemerOutputScripts`. I think we can also keep the `deriveOutputScript` function as it will be useful for future `requestRedemption` function. If we do that change, let's update the `@dev` section of docs and propagate it to `index.ts`.

Reply:
Done in https://github.com/keep-network/tbtc-v2/pull/199/commits/4b3be0a5f176ab3e5733f1876ff60fa6d3ca53a2.